File: 16. closest three sums.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>

int closest3Sums(std::vector<int> nums, int target)
{
    std::sort(nums.begin(), nums.end());
    while (nums.size() <= 3)
    {
        return std::accumulate(nums.begin(), nums.end(), 0);
    }
    int n = nums.size();
    int res = nums[0] + nums[1] + nums[2];
    for (int i = 0; i < n - 2; i++)
    {
        int j = i + 1;
        int k = n - 1;
        while (j < k)
        {
            int tmp = nums[i] + nums[j] + nums[k];
            if (abs(target - tmp) < abs(target - res))
            {
                res = tmp;
                if (res == target)
                    return res;
            }
            tmp > target ? k-- : j++;
        }
    }
    return res;
}

int main()
{
    std::vector<int> nums{11, 22, 84, 74, 23, 789, 56, -19, -49, -227, -13};
    int target = 100;
    std::cout << closest3Sums(nums, target) << std::endl;
}

File: 16. closest three sums.py
def closest3Sum(nums, target):
    nums = sorted(nums)
    res = nums[0]+nums[1]+nums[2]
    for i in range(len(nums)-2):
        j = i+1
        k = len(nums)-1
        while(j < k):
            sum = nums[i]+nums[j]+nums[k]
            if (abs(target-res) > abs(target-sum)):
                res = sum
                if(res == target):
                    return res
            if sum > target:
                k -= 1
            else:
                j += 1


if __name__ == '__main__':
    print(closest3Sum([-2, -1, 0, 1, 7, 19, 23, -47, 10], 15))
